utils/runtime-fuzzer/src/data.rs

```diff
@@ -0,0 +1,143 @@
+/// therefore to avoid printing huge amount of data we do a mock implementation of [`Debug`].
+impl Debug for FuzzerInput<'_> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_tuple("RuntimeFuzzerInput")
+            .field(&"Mock `Debug` impl")
+            .finish()
+    }
+}
+
+impl<'a> FuzzerInput<'a> {
+    pub(crate) fn inner(&self) -> &'a [u8] {
+        self.0
+    }
+
+    pub(crate) fn into_data_requirements(
+        self,
+    ) -> Result<(
+        FulfilledDataRequirement<'a, RuntimeStateViewProducer<'a>>,
+        FulfilledDataRequirement<'a, GearCallsGenerator<'a>>,
+    )> {
+        let FuzzerInput(data) = self;
+        let exec_env_data_requirement = DataRequirement::<RuntimeStateViewProducer>::new();
+        let gear_calls_data_requirement = DataRequirement::<GearCallsGenerator>::new();
+
+        let total_data_required = exec_env_data_requirement.size + gear_calls_data_requirement.size;
+        if data.len() < total_data_required {
+            log::trace!(
+                "Not enough data for fuzzing, expected - {}, got - {}",
+                total_data_required,
+                data.len(),
+            );
+
+            return Err(Error::NotEnoughData);
+        }
+
+        let (exec_env_data, gear_calls_data) = data.split_at(exec_env_data_requirement.size);
+        exec_env_data_requirement
+            .try_fulfill(exec_env_data)
+            .and_then(|eef| {
+                gear_calls_data_requirement
+                    .try_fulfill(gear_calls_data)
+                    .map(|gcf| (eef, gcf))
+            })
+    }
+}
+
+pub(crate) struct DataRequirement<T> {
+    pub(crate) size: usize,
+    _phantom: PhantomData<T>,
+}
+
+impl DataRequirement<RuntimeStateViewProducer<'_>> {
+    fn new() -> Self {
+        Self {
+            size: RuntimeStateViewProducer::random_data_requirement(),
+            _phantom: PhantomData,
+        }
+    }
+}
+
+impl DataRequirement<GearCallsGenerator<'_>> {
+    fn new() -> Self {
+        Self {
+            // Take 90% from required, because required is counted with max
+            // possible payload and salt sizes.
+            size: GearCallsGenerator::random_data_requirement() * 90 / 100,
+            _phantom: PhantomData,
+        }
+    }
+}
+
+impl<T> DataRequirement<T> {
+    pub(crate) fn try_fulfill<'a>(
+        &self,
+        data: &'a [u8],
+    ) -> Result<FulfilledDataRequirement<'a, T>> {
+        if data.len() < self.size {
+            log::trace!(
+                "Insufficient data for {:?}: expected - {}, got - {}.",
+                any::type_name::<T>(),
+                self.size,
+                data.len()
+            );
+
+            return Err(Error::NotEnoughData);
+        }
+
+        Ok(FulfilledDataRequirement {
+            data: &data[..self.size],
+            _phantom: PhantomData,
+        })
+    }
+}
+
+pub(crate) struct FulfilledDataRequirement<'a, T> {
```

Comment:
This type looks like it can be replaced by `&[u8; N]`, providing the same type-level guarantees. Then you can just store `size` as associated constant in `DataRequirement` and set `N` based on this constant

Reply:
Cool idea, but I need to have `>= N`, not `== N` data. It's impossible then to use `&[u8; N]`, isn't it? 

Also more, I'm ok with `size` being a variable

Reply:
But why? We have upper bounds of data that may be used everywhere, so overflowing this limit should be a bug, signalizing that we need to adjust this upper bound estimation

Reply:
it's a lower bound